### ISSUE TITLE
Compilation error with HIOP 0.4.6

### ISSUE THREAD

**Issue by e-aranda** [mfem/mfem#2503]
I am trying to build mfem 4.3 with hiop 0.4.6 and I have found the error

```
linalg/hiop.cpp: In member function ‘virtual void mfem::HiopNlpOptimizer::SetOptimizationProblem(const mfem::OptimizationProblem&)’:
linalg/hiop.cpp:282:61: error: invalid new-expression of abstract class type ‘mfem::HiopOptimizationProblem’
  282 |    hiop_problem = new HiopOptimizationProblem(comm, *problem);
      |                                                             ^
```
It is the same as #2204, but that error was solved in hiop 0.4.1

**Comment by cnpetra** [comment]
noted, will take a look. Do you happen to have the entire error description? Which pure virtual methods in the base class are not implemented?

**Comment by e-aranda** [comment]
The console output:

```
linalg/hiop.cpp: In member function ‘virtual void mfem::HiopNlpOptimizer::SetOptimizationProblem(const mfem::OptimizationProblem&)’:
linalg/hiop.cpp:282:61: error: invalid new-expression of abstract class type ‘mfem::HiopOptimizationProblem’
  282 |    hiop_problem = new HiopOptimizationProblem(comm, *problem);
      |                                                             ^
In file included from linalg/hiop.cpp:13:
linalg/hiop.hpp:32:7: note:   because the following virtual functions are pure within ‘mfem::HiopOptimizationProblem’:
   32 | class HiopOptimizationProblem : public hiop::hiopInterfaceDenseConstraints
      |       ^~~~~~~~~~~~~~~~~~~~~~~
In file included from linalg/hiop.hpp:25,
                 from linalg/hiop.cpp:13:
./../hiop/include/hiopInterface.hpp:127:16: note: 	‘virtual bool hiop::hiopInterfaceBase::get_prob_sizes(hiop::size_type&, hiop::size_type&)’
  127 |   virtual bool get_prob_sizes(size_type& n, size_type& m)=0;
      |                ^~~~~~~~~~~~~~
./../hiop/include/hiopInterface.hpp:139:16: note: 	‘virtual bool hiop::hiopInterfaceBase::get_vars_info(const size_type&, double*, double*, hiop::hiopInterfaceBase::NonlinearityType*)’
  139 |   virtual bool get_vars_info(const size_type& n, double *xlow, double* xupp, NonlinearityType* type)=0;
      |                ^~~~~~~~~~~~~
./../hiop/include/hiopInterface.hpp:151:16: note: 	‘virtual bool hiop::hiopInterfaceBase::get_cons_info(const size_type&, double*, double*, hiop::hiopInterfaceBase::NonlinearityType*)’
  151 |   virtual bool get_cons_info(const size_type& m, double* clow, double* cupp, NonlinearityType* type)=0;
      |                ^~~~~~~~~~~~~
./../hiop/include/hiopInterface.hpp:164:16: note: 	‘virtual bool hiop::hiopInterfaceBase::eval_f(const size_type&, const double*, bool, double&)’
  164 |   virtual bool eval_f(const size_type& n, const double* x, bool new_x, double& obj_value)=0;
      |                ^~~~~~
./../hiop/include/hiopInterface.hpp:177:16: note: 	‘virtual bool hiop::hiopInterfaceBase::eval_grad_f(const size_type&, const double*, bool, double*)’
  177 |   virtual bool eval_grad_f(const size_type& n, const double* x, bool new_x, double* gradf)=0;
      |                ^~~~~~~~~~~
./../hiop/include/hiopInterface.hpp:195:16: note: 	‘virtual bool hiop::hiopInterfaceBase::eval_cons(const size_type&, const size_type&, const size_type&, const index_type*, const double*, bool, double*)’
  195 |   virtual bool eval_cons(const size_type& n,
      |                ^~~~~~~~~
./../hiop/include/hiopInterface.hpp:422:16: note: 	‘virtual bool hiop::hiopInterfaceDenseConstraints::eval_Jac_cons(const size_type&, const size_type&, const size_type&, const index_type*, const double*, bool, double*)’
  422 |   virtual bool eval_Jac_cons(const size_type& n,
      |                ^~~~~~~~~~~~~
make[1]: *** [makefile:437: linalg/hiop.o] Error 1
make[1]: *** Se espera a que terminen otras tareas....
In function ‘void mfem::SmemPAHcurlL2Apply3D(int, int, int, int, const mfem::Array<double>&, const mfem::Array<double>&, const mfem::Array<double>&, const mfem::Vector&, const mfem::Vector&, mfem::Vector&) [with int MAX_D1D = 2; int MAX_Q1D = 3]’:
cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
In function ‘void mfem::SmemPACurlCurlApply3D(int, int, bool, int, const mfem::Array<double>&, const mfem::Array<double>&, const mfem::Array<double>&, const mfem::Array<double>&, const mfem::Array<double>&, const mfem::Array<double>&, const mfem::Vector&, const mfem::Vector&, mfem::Vector&) [with int MAX_D1D = 2; int MAX_Q1D = 3]’:
cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
```

**Comment by cnpetra** [comment]
> The console output:
> 
> ```
> linalg/hiop.cpp: In member function ‘virtual void mfem::HiopNlpOptimizer::SetOptimizationProblem(const mfem::OptimizationProblem&)’:
> linalg/hiop.cpp:282:61: error: invalid new-expression of abstract class type ‘mfem::HiopOptimizationProblem’
>   282 |    hiop_problem = new HiopOptimizationProblem(comm, *problem);
>       |                                                             ^
> In file included from linalg/hiop.cpp:13:
> linalg/hiop.hpp:32:7: note:   because the following virtual functions are pure within ‘mfem::HiopOptimizationProblem’:
>    32 | class HiopOptimizationProblem : public hiop::hiopInterfaceDenseConstraints
>       |       ^~~~~~~~~~~~~~~~~~~~~~~
> In file included from linalg/hiop.hpp:25,
>                  from linalg/hiop.cpp:13:
> ./../hiop/include/hiopInterface.hpp:127:16: note: 	‘virtual bool hiop::hiopInterfaceBase::get_prob_sizes(hiop::size_type&, hiop::size_type&)’
>   127 |   virtual bool get_prob_sizes(size_type& n, size_type& m)=0;
>       |                ^~~~~~~~~~~~~~
> ./../hiop/include/hiopInterface.hpp:139:16: note: 	‘virtual bool hiop::hiopInterfaceBase::get_vars_info(const size_type&, double*, double*, hiop::hiopInterfaceBase::NonlinearityType*)’
>   139 |   virtual bool get_vars_info(const size_type& n, double *xlow, double* xupp, NonlinearityType* type)=0;
>       |                ^~~~~~~~~~~~~
> ./../hiop/include/hiopInterface.hpp:151:16: note: 	‘virtual bool hiop::hiopInterfaceBase::get_cons_info(const size_type&, double*, double*, hiop::hiopInterfaceBase::NonlinearityType*)’
>   151 |   virtual bool get_cons_info(const size_type& m, double* clow, double* cupp, NonlinearityType* type)=0;
>       |                ^~~~~~~~~~~~~
> ./../hiop/include/hiopInterface.hpp:164:16: note: 	‘virtual bool hiop::hiopInterfaceBase::eval_f(const size_type&, const double*, bool, double&)’
>   164 |   virtual bool eval_f(const size_type& n, const double* x, bool new_x, double& obj_value)=0;
>       |                ^~~~~~
> ./../hiop/include/hiopInterface.hpp:177:16: note: 	‘virtual bool hiop::hiopInterfaceBase::eval_grad_f(const size_type&, const double*, bool, double*)’
>   177 |   virtual bool eval_grad_f(const size_type& n, const double* x, bool new_x, double* gradf)=0;
>       |                ^~~~~~~~~~~
> ./../hiop/include/hiopInterface.hpp:195:16: note: 	‘virtual bool hiop::hiopInterfaceBase::eval_cons(const size_type&, const size_type&, const size_type&, const index_type*, const double*, bool, double*)’
>   195 |   virtual bool eval_cons(const size_type& n,
>       |                ^~~~~~~~~
> ./../hiop/include/hiopInterface.hpp:422:16: note: 	‘virtual bool hiop::hiopInterfaceDenseConstraints::eval_Jac_cons(const size_type&, const size_type&, const size_type&, const index_type*, const double*, bool, double*)’
>   422 |   virtual bool eval_Jac_cons(const size_type& n,
>       |                ^~~~~~~~~~~~~
> make[1]: *** [makefile:437: linalg/hiop.o] Error 1
> make[1]: *** Se espera a que terminen otras tareas....
> In function ‘void mfem::SmemPAHcurlL2Apply3D(int, int, int, int, const mfem::Array<double>&, const mfem::Array<double>&, const mfem::Array<double>&, const mfem::Vector&, const mfem::Vector&, mfem::Vector&) [with int MAX_D1D = 2; int MAX_Q1D = 3]’:
> cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
> cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
> cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
> cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
> cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
> In function ‘void mfem::SmemPACurlCurlApply3D(int, int, bool, int, const mfem::Array<double>&, const mfem::Array<double>&, const mfem::Array<double>&, const mfem::Array<double>&, const mfem::Array<double>&, const mfem::Array<double>&, const mfem::Vector&, const mfem::Vector&, mfem::Vector&) [with int MAX_D1D = 2; int MAX_Q1D = 3]’:
> cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
> cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
> cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
> cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
> cc1plus: warning: ‘void* __builtin_memcpy(void*, const void*, long unsigned int)’ writing between 8 and 34359738368 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
> ```

got it thanks. Just some ordinal type mismatches caused by a recent change in the interface

**Comment by tzanio** [comment]
Discussion continued in #2517

**Comment by e-aranda** [comment]
Hello, mfem 4.3 with hiop 0.5.3 is giving the same error as before. I thought it was already corrected... or maybe I am doing something wrong?

**Comment by tzanio** [comment]
This PR was merged after the 4.3 release (and will be part of mfem-4.4). Can you try mfem:master instead?

**Comment by e-aranda** [comment]
Ok, it works fine with mfem::master. Thanks.
